proofs/alf/cvc5/programs/Strings.smt3

```diff
@@ -702,10 +711,223 @@
+  (($str_re_includes_rec (re.++ (str.to_re "") r2) r3)              ($str_re_includes_rec r2 r3))
+  (($str_re_includes_rec r1 (re.++ (str.to_re (str.++ s1 s2)) r4))  ($str_re_includes_rec r1 (re.++ (str.to_re s1) (str.to_re s2) r4)))
+  (($str_re_includes_rec r1 (re.++ (str.to_re "") r4))              ($str_re_includes_rec r1 r4))
+  ; otherwise we recurse for three possible cases
+  (($str_re_includes_rec (re.++ r1 r2) (re.++ r3 r4))
+    (alf.ite (alf.ite ($str_re_includes ($re_from_flat_form r1 false) ($re_from_flat_form r3 false)) ; can skip both if we have a base inclusion
+               ($str_re_includes_rec r2 r4)
+               false)
+      true
+    (alf.ite (alf.ite ($re_is_unbound_wildcard (re.++ r1 r2)) ; can skip LHS if unbounded wildcard
+              ($str_re_includes_rec r2 (re.++ r3 r4))
+              false)
+      true
+    (alf.ite (alf.ite (alf.is_eq r1 (re.* re.allchar))         ; can skip RHS if it is sigma star
+              ($str_re_includes_rec (re.++ r1 r2) r4)
+              false)
+      true
+      false))))
+  ; handles the case where there is a leftover (re.* re.allchar) on the LHS
+  (($str_re_includes_rec (re.++ r1 r2) @re.empty)       (alf.and (alf.is_eq r1 (re.* re.allchar)) (alf.is_eq r2 @re.empty)))
+  (($str_re_includes_rec r1 r3)                         false)
+  )
+)
+
+; program: $str_re_includes
+; args:
+; - r1 RegLan: The first regular expression.
+; - r2 RegLan: The second regular expression.
+; return: >
+;   true if we can infer that all strings in r2 are also in r1.
+(program $str_re_includes ((s1 String) (s2 String) (s3 String) (s4 String) (r1 RegLan) (r2 RegLan :list) (r3 RegLan))
+  (RegLan RegLan) Bool
+  (
+  (($str_re_includes r1 r1)                               true)
+  (($str_re_includes r1 (str.to_re s1))                   ($str_eval_str_in_re s1 r1))
+  (($str_re_includes (str.to_re s1) r1)                   false)
+  (($str_re_includes (re.range s1 s2) (re.range s3 s4))   (let ((z1 (alf.to_z s1)))
+                                                          (let ((z2 (alf.to_z s2)))
+                                                          (let ((z3 (alf.to_z s3)))
+                                                          (let ((z4 (alf.to_z s4)))
+                                                          (alf.requires (alf.is_neg z1) false
+                                                          (alf.requires (alf.is_neg z2) false
+                                                          (alf.requires (alf.is_neg z3) false
+                                                          (alf.requires (alf.is_neg z4) false
+                                                            (alf.and ($compare_geq z2 z3) ($compare_geq z3 z1)
+                                                                  ($compare_geq z2 z4) ($compare_geq z4 z1)))))))))))
```

Comment:
This is making the assumption that ranges are non-empty. Otherwise the re.range would be rewritten.

I think its probably easier to follow the internals for now, as it makes things simpler to check (despite handling fewer cases).